File: 41/41_version1.java
// version 2: partition & hashcode
class Solution {
    private static int DEFAULTANS = 1;
    public int firstMissingPositive(int[] nums) {
        int posNum = partition(nums) + 1;
        if(posNum==0) return DEFAULTANS;
        int index = 0;
        while(index < posNum){
            int value = Math.abs(nums[index]);
            if(value<=posNum && nums[value-1] > 0) nums[value-1] = -nums[value-1];
            index+=1;
        }
        index = 0;
        while(index < posNum){
            if(nums[index] > 0) return index+1;
            index+=1;
        }
        return posNum+1;
    }
    
    private int partition(int[] nums){
        int index = 0;
        int position = -1;
        while(index < nums.length){
            if(nums[index] > 0){
                position+=1;
                swap(nums,position,index);
            }
            index+=1;
        }
        return position;
    }
    
    private void swap(int[] nums,int index1,int index2){
        if(index1!=index2){
            int temp = nums[index1];
            nums[index1] = nums[index2];
            nums[index2] = temp;
        }
    }
}

File: 41/41_version1.py
# version 2: partition
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        # as now we have [pos_num] positive elements
        # the first missing element must exists in range[1,pos_num]
        pos_num = self.partition(nums)+1
        # corner case: there is no positive number in nums array
        if pos_num == 0:
            return 1
        
        # step 2: use nums[i] to stand for if i+1 exists(keep original information of nums)
        # i>=0 && i<k
        index = 0
        while index < pos_num:
            value = abs(nums[index])
            if value <= pos_num and nums[value-1] > 0:
                nums[value-1] = -nums[value-1]
            index+=1
        index = 0
        while index < pos_num:
            if nums[index] > 0:
                return index+1
            index+=1
        
        return pos_num+1
    
    
    def partition(self,nums):
        """
        partition the nums array, put all positive elements at the left side
        return: the end index of last positive element
        """
        index = 0
        pos=-1
        while index < len(nums):
            if nums[index] > 0:
                pos+=1
                if pos!=index:
                    nums[index],nums[pos] = nums[pos],nums[index]
            index+=1
        return pos
